fifo.go

```diff
@@ -204,6 +204,10 @@ func (f *fifo) Write(b []byte) (int, error) {
 // before open(2) has returned and fifo was never opened.
 func (f *fifo) Close() (retErr error) {
 	for {
+		if f == nil {
+			return
```

Comment:
Also not sure if we should 

- return an error in this case (currently it would use `retErr`, if set)
- return `nil` (which would unset `retErr` if it was set before)

Reply:
I think the caller should check for nil.

Reply:
Could it _become_ `nil` while we're in this `for{}` loop?

Reply:
I should hope not... if so then that would be the bug that should be fixed.

Reply:
Hmm... so I suspect that something like that actually happens (somehow); following that trace, the panic followed from this code; https://github.com/containerd/containerd/blob/269548fa27e0089a8b8278fc4fc781d7f65a939b/cio/io.go#L199-L206

```go
func (c *cio) Close() error {
	var lastErr error
	for _, closer := range c.closers {
		if closer == nil {
			continue
		}
		if err := closer.Close(); err != nil {
			lastErr = err
		}
	}
	return lastErr
}
```

So there's already a `nil` check there. So only other explanation would be there there's some concurrency happening, in which case do we need locking?

Reply:
could have been explain by the infamous interface is nil problem, either we make sure that we dont pass nil io.Closer to cio, or do a type check ( yuck), or we check inside fifo.Close for nil.

Reply:
Ah, sorry, missed your comment here, @dqminh - do you mean we should use a different fix, or is the current "fix" ok?

Reply:
i think the best case is that we need to make sure the caller doesn't pass a nil interface as that can be nasty too if we call anything other than Close ( which is fixed in this way ). The current check we are doing here is not the end of the world, but ideally i don't think we need it at this place.